File: data-pipeline/pipeline/engine/separation.py
"""Low-intensity magnetic separation and desliming.

LIMS: liberated magnetite is captured with ``p_max (1 - exp(-d/d_f))``; composites with a capture
that rises with their magnetite content, ``p_max (1 - exp(-c/c0)) (1 - exp(-d/d_f))``; free gangue
reports by entrapment ``e0 + e1 exp(-d/d_e)``, scaled down in the cleaner drum. Concentrate Fe
grade therefore follows liberation, and so the grind (Muthaphuli 2014, JSAIMM 114(7)).
Desliming: a cyclone at a fine cut sends slimes to tailings with a water bypass to the underflow.
Sources: docs/frameworks/magnetic-separation and docs/frameworks/desliming.
"""
from __future__ import annotations

import math
from dataclasses import dataclass

import numpy as np

from .cyclone import corrected_cut, reduced_partition
from .grid import grid
from .model import DeslimePlant, MagneticPlant
from .ore import ResolvedOre
from .species import SpeciesDef, partition_species, species_defs, to_minerals
from .streams import Stream


def lims_capture(defs: list[SpeciesDef], ore: ResolvedOre, mp: MagneticPlant, entrapment_factor: float) -> dict[str, np.ndarray]:
    size = grid().size
    fine = 1.0 - np.exp(-size / mp.fine_scale_um)
    entrapment = entrapment_factor * (mp.entrapment_base + mp.entrapment_fines * np.exp(-size / mp.entrapment_scale_um))
    out: dict[str, np.ndarray] = {}
    for d in defs:
        magnetic = ore.spec[d.mineral].magnetic
        if magnetic and d.kind == "liberated":
            out[d.id] = mp.max_capture * fine
        elif magnetic and d.kind == "composite":
            c = ore.spec[d.mineral].composite_content
            out[d.id] = mp.max_capture * (1.0 - math.exp(-c / mp.composite_threshold)) * fine
        else:
            out[d.id] = entrapment
    return out


@dataclass
class MagneticResult:
    streams: dict[str, Stream]
    capture_rougher: dict[str, np.ndarray]
    capture_cleaner: dict[str, np.ndarray]
    defs: list[SpeciesDef]


def _split_water(mags_solids: float, feed_water: float, solids_fraction: float) -> float:
    return min(feed_water, mags_solids * (1.0 - solids_fraction) / solids_fraction)


def run_magnetic(x: dict[str, np.ndarray], feed_water: float, ore: ResolvedOre, mp: MagneticPlant) -> MagneticResult:
    defs = species_defs(ore)
    cap_r = lims_capture(defs, ore, mp, 1.0)
    cap_c = lims_capture(defs, ore, mp, mp.cleaner_factor)
    mags_r = {k: cap_r[k] * x[k] for k in x}
    tail_r = {k: x[k] - mags_r[k] for k in x}
    water_mags_r = _split_water(sum(float(np.sum(v)) for v in mags_r.values()), feed_water, mp.concentrate_solids)
    mags_c = {k: cap_c[k] * mags_r[k] for k in x}
    tail_c = {k: mags_r[k] - mags_c[k] for k in x}
    water_mags_c = _split_water(sum(float(np.sum(v)) for v in mags_c.values()), water_mags_r, mp.concentrate_solids)

    def stream(species: dict[str, np.ndarray], water: float) -> Stream:
        return Stream(to_minerals(species, defs, ore), water)

    return MagneticResult({
        "lims_feed": stream(x, feed_water),
        "lims_rougher_concentrate": stream(mags_r, water_mags_r),
        "lims_rougher_tail": stream(tail_r, feed_water - water_mags_r),
        "lims_cleaner_concentrate": stream(mags_c, water_mags_c),
        "lims_cleaner_tail": stream(tail_c, water_mags_r - water_mags_c),
    }, cap_r, cap_c, defs)


@dataclass
class DeslimeResult:
    underflow: Stream
    slimes: Stream
    partition: dict[str, np.ndarray]
    underflow_species: dict[str, np.ndarray]
    underflow_water: float


def run_deslime(x: dict[str, np.ndarray], feed_water: float, ore: ResolvedOre, dp: DeslimePlant, cut_um: float) -> DeslimeResult:
    defs = species_defs(ore)
    rho_ref = ore.density[ore.host]
    partition: dict[str, np.ndarray] = {}
    for d in defs:
        partition[d.id] = dp.bypass + (1.0 - dp.bypass) * reduced_partition(corrected_cut(cut_um, rho_ref, d.density), dp.sharpness)
    under, over = partition_species(x, partition)
    water_under = dp.bypass * feed_water
    return DeslimeResult(Stream(to_minerals(under, defs, ore), water_under),
                         Stream(to_minerals(over, defs, ore), feed_water - water_under), partition, under, water_under)

File: frontend/src/engine/separation.ts
/**
 * Low-intensity magnetic separation and desliming (port of engine/separation.py; theory in
 * docs/methodologies/07_magnetic-separation.md and 08_desliming.md).
 */
import { correctedCut, reducedPartition } from './cyclone';
import { grid, type Vec } from './grid';
import type { DeslimePlant, MagneticPlant } from './model';
import type { ResolvedOre } from './ore';
import { partitionSpecies, speciesDefs, toMinerals, type Species, type SpeciesDef } from './species';
import { Stream } from './streams';

export function limsCapture(defs: SpeciesDef[], ore: ResolvedOre, mp: MagneticPlant, entrapmentFactor: number): Species {
  const size = grid().size;
  const n = size.length;
  const fine = new Float64Array(n);
  const entrapment = new Float64Array(n);
  for (let i = 0; i < n; i += 1) {
    fine[i] = 1.0 - Math.exp(-size[i] / mp.fine_scale_um);
    entrapment[i] = entrapmentFactor * (mp.entrapment_base + mp.entrapment_fines * Math.exp(-size[i] / mp.entrapment_scale_um));
  }
  const out: Species = {};
  for (const d of defs) {
    const magnetic = ore.spec[d.mineral].magnetic;
    const v = new Float64Array(n);
    if (magnetic && d.kind === 'liberated') {
      for (let i = 0; i < n; i += 1) v[i] = mp.max_capture * fine[i];
    } else if (magnetic && d.kind === 'composite') {
      const c = ore.spec[d.mineral].composite_content;
      const scale = mp.max_capture * (1.0 - Math.exp(-c / mp.composite_threshold));
      for (let i = 0; i < n; i += 1) v[i] = scale * fine[i];
    } else {
      v.set(entrapment);
    }
    out[d.id] = v;
  }
  return out;
}

export type MagneticResult = { streams: Record<string, Stream>; capture_rougher: Species; capture_cleaner: Species; defs: SpeciesDef[] };

function splitWater(magsSolids: number, feedWater: number, solidsFraction: number): number {
  return Math.min(feedWater, magsSolids * (1.0 - solidsFraction) / solidsFraction);
}

const total = (species: Species) => { let s = 0.0; for (const v of Object.values(species)) for (let i = 0; i < v.length; i += 1) s += v[i]; return s; };

export function runMagnetic(x: Species, feedWater: number, ore: ResolvedOre, mp: MagneticPlant): MagneticResult {
  const defs = speciesDefs(ore);
  const capR = limsCapture(defs, ore, mp, 1.0);
  const capC = limsCapture(defs, ore, mp, mp.cleaner_factor);
  const magsR: Species = {}; const tailR: Species = {}; const magsC: Species = {}; const tailC: Species = {};
  for (const k of Object.keys(x)) {
    const n = x[k].length;
    magsR[k] = new Float64Array(n); tailR[k] = new Float64Array(n); magsC[k] = new Float64Array(n); tailC[k] = new Float64Array(n);
    for (let i = 0; i < n; i += 1) {
      magsR[k][i] = capR[k][i] * x[k][i];
      tailR[k][i] = x[k][i] - magsR[k][i];
    }
  }
  const waterMagsR = splitWater(total(magsR), feedWater, mp.concentrate_solids);
  for (const k of Object.keys(x)) {
    for (let i = 0; i < x[k].length; i += 1) {
      magsC[k][i] = capC[k][i] * magsR[k][i];
      tailC[k][i] = magsR[k][i] - magsC[k][i];
    }
  }
  const waterMagsC = splitWater(total(magsC), waterMagsR, mp.concentrate_solids);
  const stream = (species: Species, water: number) => new Stream(toMinerals(species, defs, ore), water);
  return {
    streams: {
      lims_feed: stream(x, feedWater),
      lims_rougher_concentrate: stream(magsR, waterMagsR),
      lims_rougher_tail: stream(tailR, feedWater - waterMagsR),
      lims_cleaner_concentrate: stream(magsC, waterMagsC),
      lims_cleaner_tail: stream(tailC, waterMagsR - waterMagsC),
    },
    capture_rougher: capR, capture_cleaner: capC, defs,
  };
}

export type DeslimeResult = { underflow: Stream; slimes: Stream; partition: Species; underflow_species: Species; underflow_water: number };

export function runDeslime(x: Species, feedWater: number, ore: ResolvedOre, dp: DeslimePlant, cutUm: number): DeslimeResult {
  const defs = speciesDefs(ore);
  const rhoRef = ore.density[ore.host];
  const partition: Species = {};
  for (const d of defs) {
    const y = reducedPartition(correctedCut(cutUm, rhoRef, d.density), dp.sharpness);
    const v: Vec = new Float64Array(y.length);
    for (let i = 0; i < y.length; i += 1) v[i] = dp.bypass + (1.0 - dp.bypass) * y[i];
    partition[d.id] = v;
  }
  const [under, over] = partitionSpecies(x, partition);
  const waterUnder = dp.bypass * feedWater;
  return { underflow: new Stream(toMinerals(under, defs, ore), waterUnder), slimes: new Stream(toMinerals(over, defs, ore), feedWater - waterUnder),
    partition, underflow_species: under, underflow_water: waterUnder };
}
